### ISSUE TITLE
Seg fault and something even more insidious (memory corruption?)

### ISSUE THREAD

**Issue by christian-storm** [opencog/link-grammar#376]
After pulling my hair out for a while on this one I finally figured out what is going on.  I was processing a batch of sentences using 5.3.6 python bindings.  In order to get it to work I had to set max_parse_time to 5 seconds so that a particular sentence wouldn't segfault the link-parser.  After getting it to process them all without crashing I couldn't understand why the linkage count was going to zero on most of the sentences.  I then narrowed it down to that one sentence that required me to set the timeout.  After that sentence, I discovered that all the following sentences had zero linkages.  I wrote a modified example.py script to demonstrate (at end of issue).  When I run it I get the following output:

```
python example.py 
('Version:', 'link-grammar-5.3.6')
English: Found 18 linkages (18 had no P.P. violations)
English: Found 0 linkages (0 had no P.P. violations)  <---With timeout
English: Found 0 linkages (0 had no P.P. violations)  <---Should be same as first (18 linkages)
English: Found 2147483647 linkages (0 of 100 random linkages had no P.P. violations)  <--Without timeout
python(91714,0x7fff783be000) malloc: *** error for object 0x1049f0004: pointer being freed was not allocated
*** set a breakpoint in malloc_error_break to debug
Abort trap: 6

```
My guess is that even with the timeout, some memory holding state is being corrupted and causes all downstream processing to be affected. 

```
#! /usr/bin/env python
# -*- coding: utf8 -*-
#
# Link Grammar example usage
#
import locale

from linkgrammar import Sentence, ParseOptions, Dictionary
try:
    import linkgrammar._clinkgrammar as clg
except ImportError:
    import _clinkgrammar as clg

locale.setlocale(locale.LC_ALL, "en_US.UTF-8")

print("Version:", clg.linkgrammar_get_version())

def s(q):
    return '' if q == 1 else 's'

def linkage_stat(psent, lang):
    """
    This function mimics the linkage status report style of link-parser
    """
    random = ' of {0} random linkages'. \
             format(psent.num_linkages_post_processed()) \
             if psent.num_valid_linkages() < psent.num_linkages_found() else ''

    print('{0}: Found {1} linkage{2} ({3}{4} had no P.P. violations)'. \
          format(lang, psent.num_linkages_found(),
                 s(psent.num_linkages_found()),
                 psent.num_valid_linkages(), random))

# Parses fine 
po = ParseOptions(max_parse_time=5)
sent = Sentence("Members gather money for the funeral and help them.", Dictionary(), po)
linkages = sent.parse()
linkage_stat(sent, 'English')

# Hits timeout of 5 seconds
sent = Sentence("While the travel company will most likely show them some interesting sites in order for their customers to advertise for their company to their family and friends, it is highly unlikely, that the company will tell about the sites that were not included in the tour -- for example due to entrance fees that would make the total package price overly expensive.", Dictionary(), po)
linkages = sent.parse()
linkage_stat(sent, 'English')

# Parse error:  What happened to all the linkages?
sent = Sentence("Members gather money for the funeral and help them.", Dictionary(), po)
linkages = sent.parse()
linkage_stat(sent, 'English')

# Remove timeout to demonstrate seg fault 
po = ParseOptions()
sent = Sentence("While the travel company will most likely show them some interesting sites in order for their customers to advertise for their company to their family and friends, it is highly unlikely, that the company will tell about the sites that were not included in the tour -- for example due to entrance fees that would make the total package price overly expensive.", Dictionary(), po)
linkages = sent.parse()
linkage_stat(sent, 'English')

```

**Comment by christian-storm** [comment]
For what it's worth it doesn't seem to happen from the commandline:

```
linkparser> !graphics
Graphical display of linkage turned off.
linkparser> !timeout=5
timeout set to 5
linkparser> Members gather money for the funeral and help them.
Found 18 linkages (18 had no P.P. violations)
	Linkage 1, cost vector = (UNUSED=0 DIS= 0.00 LEN=25)
Press RETURN for the next linkage.
linkparser> While the travel company will most likely show them some interesting sites in order for their customers to advertise for their company to their family and friends, it is highly unlikely, that the company will tell about the sites that were not included in the tour -- for example due to entrance fees that would make the total package price overly expensive.
No complete linkages found.
Timer is expired!
Entering "panic" mode...
link-grammar: WARNING: Combinatorial explosion! nulls=1 cnt=2147483647
Consider retrying the parse with the max allowed disjunct cost set lower.
At the command line, use !cost-max

Found 2147483647 linkages (0 of 100 random linkages had no P.P. violations) at null count 1
linkparser> Members gather money for the funeral and help them.
Found 18 linkages (18 had no P.P. violations)
	Linkage 1, cost vector = (UNUSED=0 DIS= 0.00 LEN=25)
Press RETURN for the next linkage.
```

**Comment by linas** [comment]
Yep. Current github command line does not show any of these bugs.  Current github python bindings do show the `English: Found 0 linkages (0 had no P.P. violations)` bug for the second short sentence.  It does not show the crash for the second long sentence. Looking at the ChangeLog, I see that there was a crash fixed in 5.3.7 it was pull request #300 but it was billed as affecting only Russian. Maybe it fixes the current crash.

That leaves the mystery of why python is reporting 0 parses for the second short sentence. 

You might want to try version 5.3.9 to avoid the crash

**Comment by linas** [comment]
Err, I mean, confirm that 5.3.9 doesn't crash -- although its fine for me, things may be different on your machine/os/compiler combination.

**Comment by linas** [comment]
And the 0 parses report is because it fails to reset the timeout -- the second short parse halts immediatelty, because it thinks the timeout has already popped.  Strangely, this appears to be a very old bug, too  -- since the command-line version is not affected by it!

**Comment by christian-storm** [comment]
Unfortunately, I can't install 5.3.9 on MacOS because of a locale_t bug https://github.com/opencog/link-grammar/issues/374 but ampli says he is taking a look at it.

Starting from a random spot in a set of sentences I find that it works for 20 or so sentences and then starts continually reporting 0 afterwards.

For example:
```
num linkages	Sentence
36	Disadvantage is parking their car is very difficult.
946326	They draw the consumers, like me, to purchase this great product with all these amazing ingredients and all that but actually they just sometimes make something up just to increase their sales.
84	"This is an easy way of life, because they make only the ""secure"" things."
85344	I want to talk about nocive or bad products like alcohol, hair spray and cigarrets.
0	For example they can play football whenever they want but the olders can not.
404	if the quality of the product is good and it is worth using customer will definetly buy it.
8188	This isnt very helpfull cause the owner needs to spend more money on transport and goods for the workers.
198	Bigger farming are use more chemical product and substance to feed fish .
3680	Whereas, a community for the old and underprivleged would want help, aid and assistence from the youth.
2464	If our scintist may not take the risk then how we can develope the new things and how we can make our life very comfartable and enjoyful.
0	For example, if people wants to go traveling by plane, which will be definitely a popular transportation in the future, they have to drive a car or take a taxi or a bus to the airport.
4	Forexample, My cousin is 12years old.
35	Finally student become more knowladgable and confident about knowladge.
0	It figures Diana Krall wearing a Rolex watch and has a text that suggests that if the reader wants to belong to the restricted club, not necessarily of Rolex watch wearers, but of highly talented, sophisticated and competent artists, he or she should also have a Rolex.
0	When we started evaluating on those to drop and those not to, I had dropped about a lot and now I can concetrate much better with the little I offer.
1072	if it does, you will only get some misundrestood idea about that acticle but not the original one.
3880	Secondly, chimp can also present the capability for grammer which means it can demonstrate the ability to combine words and be able to use grammatical constructions.
6	Nevertheless, this city is having a big change.
0	I think to learn facts is only an evidence if the students can work hard and if they are able to keep every detailled information as tought, but it's definitly not an evidence for their intelligence.
0	Unforturntly, almost older people can not use internet, in spite of benefit of internet.
4	Facts can learn through many books.
0	So, if i have alot of information about this subject , i will taulk too much with knowledge but if i have general information for this subject , i will talk about this subjec with my limited knowlege and this case may be make me shame like when my brother asked me about some thing but i have not alot of information about this thing.
0	becuse if i see some one did somthing to may safe me time and engy and it wok's i will do it .
0	But when i finished medical school i was only able to specialize in one field because you can never master all what you have studied in medical school, which makes sense in a way or else there would be no specialities in medicine.
0	Maybe some of them just couldnt graduate from a single academic course but this is because he is not that type of human.
0	He/she is simply left with no time to cope with mundane things, leave alone the community he/she lives in!
0	There are several reason.
0	According to the reading the dinosaur is not an endotherm.
0	The government should think about the whole nature.

```
At a minimum, I know that the last sentence has 32 linkages via the command line.

**Comment by linas** [comment]
There is no python binding to reset the resource allocation.  You can work around this with the following path, and recompiling:
```
diff --git a/link-grammar/api.c b/link-grammar/api.c
index 2ec8157..387c2b8 100644
--- a/link-grammar/api.c
+++ b/link-grammar/api.c
@@ -1503,10 +1508,7 @@ int sentence_parse(Sentence sent, Parse_Options opts)
 
        /* Initialize/free any leftover garbage */
        free_sentence_disjuncts(sent);  /* Is this really needed ??? */
-       resources_reset_space(opts->resources);
-
-       if (resources_exhausted(opts->resources))
-               return 0;
+       resources_reset(opts->resources);
 
        /* Expressions were previously set up during the tokenize stage. */
        expression_prune(sent);
````

**Comment by christian-storm** [comment]
I'll try that...I found another potential cause as well- combinatorial explosion.

It is true that consumers preffer to buy a product that has a lower price, but when international companies that has already the certification begins to send its market , people will preffer to consume theirs because the difference between prices is probbably not going to affect them to much.
link-grammar: WARNING: Combinatorial explosion! nulls=0 cnt=19852576
Consider retrying the parse with the max allowed disjunct cost set lower.
At the command line, use !cost-max

**Comment by linas** [comment]
Combinatorial explosions are not uncommon for long sentences: at various points in the sentence, there are 2 or 3 or 8 difference choices, and multiplying them together can give a very large number.  It also doesn't help when there are typos and grammar issues:  "preffer" (twice) "probbably"  "companies that has"  ... "has already the certification" which typically increase the number of parse choices -- although fixing these does not really affect the final result very much.

I've got some ideas on how to handle the choice explosion more elegantly, but doing so has been very low priority.

**Comment by linas** [comment]
Anyway, pull req #377 fixes the zero-linkages problem, and I am guessing that pull req #300 should fix the crash.

Both of these fixes are very tiny, and you should be able to apply them by hand.  I am very interested in finding out if the system still crashes, after these two fixes.

**Comment by christian-storm** [comment]
I patched the code and I am happy to report that it seems to have fixed both issues.  Thanks Linas for jumping on it.

**Comment by linas** [comment]
Welcome!  Closing this, the fixes will be in version 5.3.10
